cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/MaintenancePlan.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2020 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License"). See License in the project root for license information.
+ */
+
+package com.linkedin.kafka.cruisecontrol.detector;
+
+import java.nio.ByteBuffer;
+
+
+/**
+ * An abstract class for all maintenance plans.
+ */
+public abstract class MaintenancePlan {
+  protected final MaintenanceEventType _maintenanceEventType;
+  protected final long _timeMs;
+  protected final int _brokerId;
+
+  public MaintenancePlan(MaintenanceEventType maintenanceEventType, long timeMs, int brokerId) {
+    _maintenanceEventType = maintenanceEventType;
+    _timeMs = timeMs;
+    _brokerId = brokerId;
+  }
+
+  /**
+   * Retrieve the maintenance event type of this maintenance plan.
+   *
+   * @return the maintenance event type of this maintenance plan, which is stored in the serialized metrics so
+   * that the deserializer will know which class should be used to deserialize the data.
+   */
+  public MaintenanceEventType maintenanceEventType() {
+    return _maintenanceEventType;
+  }
+
+  /**
+   * @return The timestamp in ms that corresponds to the generation of this plan.
+   */
+  public long timeMs() {
+    return _timeMs;
+  }
+
+  /**
+   * @return The id of the broker that reported this maintenance event.
+   */
+  public int brokerId() {
+    return _brokerId;
+  }
+
+  /**
+   * @return The plan version for serialization/deserialization.
+   */
+  public abstract byte planVersion();
+
+  /**
+   * Serialize the maintenance plan to a byte buffer with the header size reserved.
+   *
+   * @param headerSize the header size to reserve.
+   * @return A {@link ByteBuffer} with header size reserved at the beginning.
+   */
+  public ByteBuffer toBuffer(int headerSize) {
```

Comment:
Seems like this protocol is going to have many different types of messages, but low QPS.  Seems like a good use case for some kind of serialization library.

Reply:
Can you comment a little more on the advantages of using a serialization library as opposed to the current approach?
I feel that we may not get the same level of efficiency and may need to require users to have an extra dependency on their end -- and potentially on our end depending on the selected library.

Reply:
You probably won't gain any efficiency.  Maybe protobuf since the integers are variable length encoded.  You gain maintainability.   Since this is low qps there's no need for efficiency here.  Gson looks like it is easy to use and has minimal dependencies. The real win is not when you write the initial code it's when you need to change the code: add additional messages, add additional fields, need to reason about versioning, need to debug messages by actually reading their contents.

Reply:
Makes sense -- updated the code to use Gson for de/serialization of maintenance plans.